Executable and library rust project types

Undin opened the pull request:
Closes #1277 

Add checkbox into new project wizard to choose project/module type: executable or library.

Undin commented on src/main/resources/META-INF/idea-only.xml:
```diff
@@ -1,6 +1,5 @@
 <idea-plugin version="2">
     <extensions defaultExtensionNs="com.intellij">
-        <moduleBuilder builderClass="org.rust.ide.idea.RsModuleBuilder"/>
```
Do we really need this? I dropped it because `RsModuleBuilder` has not empty constructor now and IDEA fails to create it with reflection but I don't know the purpose of this extension (kotlin plugin doesn't have it too).
For me, the plugin works well without it:)

Undin commented on src/main/kotlin/org/rust/ide/idea/RsModuleBuilder.kt:
```diff
@@ -9,29 +9,34 @@ import com.intellij.execution.ExecutionException
-            Disposer.register(parentDisposable, Disposable { this.disposeUIResources() })
-        }
+    override fun createWizardSteps(wizardContext: WizardContext, modulesProvider: ModulesProvider): Array<ModuleWizardStep> =
+        arrayOf(CargoConfigurationWizardStep(wizardContext))
```
I'm not sure how to register `Disposable` here because IDEA doesn't pass `parentDisposable` into `createWizardSteps` method.

matklad replied:
Frankly, I am not even sure we need to call `disposeUIresources` at all.

matklad commented on src/main/kotlin/org/rust/cargo/project/settings/ui/RustProjectSettingsPanel.kt:
```diff
@@ -94,11 +94,11 @@ class RustProjectSettingsPanel(private val cargoProjectDir: String = ".") : Disp
+            autoUpdateEnabled = true,
+            explicitPathToStdlib = null,
+            useCargoCheckForBuild = false,
+            useCargoCheckAnnotator = true
```
By the way, it is absolutely horrendous that there is this super boring similar code in like six places! I definitely over engineered it. If you have any ideas on how to remove the duplication, please go ahead!

matklad commented:
The implementation looks great, however, I have a few questions about the general UI structure :) 

I am not really a fan of multistep wizards, and I in general I think the better approach is "create something runnable in a single click, and make it easy to change it later". So I would prefer if instead of this separate template step, we just had a `--bin` checkbox on the single config step we have now.

It might be rather challenging to add this checkbox, because we try to share the same form in several places. However, perhaps this is just wrong? I think the following options are just useless on this dialog, because they don't affect creation of the project in any way

![new-project](https://user-images.githubusercontent.com/1711539/27725077-6d90a12e-5d7d-11e7-9cc0-8f5a6aee94bd.png)

So what do you thing about removing a reusable `RustProjectSettingsPanel` altogether, and building a custom UI for project creation step, reusing only some fields?

matklad commented:
>make it easy to change it later"

I think we have quite a few opportunities here to add some inspections/intentions to `Cargo.toml` file, like "add binary", "add library", "warning, file `foo.rs` does not belong to any crate, do you want to create a foo binary"

Undin commented:
Do you really want to remove "checkbox" settings from new project wizard and keep them only in project settings panel? Yes, these settings don't affect the creation of the project but they are project settings too and I think it can be annoying to change some setting(s) for every new project if you don't like default value.

Undin commented:
I dropped template wizard step and added checkbox for executable project creation.
I didn't remove other checkboxes because I'm not sure it's good decision but if you consider so, I'll move them from `RustProjectSettingsPanel` into `RustProjectConfigurable`

matklad commented:
>Do you really want to remove "checkbox" settings from new project wizard and keep them only in project settings panel?

Yeah I still think we don't need those checkboxes.

> it can be annoying to change some setting(s) for every new project if you don't like default value.

There's a little known feature that you can configure default settings for the project :) `Ctrl+Shift+A > Default Settings`. For non default settings, I think it doesn't make much difference if you configure them on creating a project, or afterwards via settings. 


That said, 

bors r+

because changing UI here is a separate task.

matklad commented:
Ah, one more thing though: let's keep `--bin` as the default? I think it makes more sense, and actually we want to switch to `--bin` by default in Cargo as well!

matklad commented:
(test failure is a spurious network error)

Undin commented:
> Yeah I still think we don't need those checkboxes.
Ok, I can remove them from new project wizard in next PR:)

Undin commented:
> Ah, one more thing though: let's keep --bin as the default? I think it makes more sense, and actually we want to switch to --bin by default in Cargo as well!

No problem:)
Done!

matklad commented:
bors r+

matklad commented:
Thanks!
